### PR TITLE
Removes redundant outers

### PR DESCRIPTION
**Pull request by odersky**
Widens the criterion when outer fields can be omitted. It used to be that sub- and superclass had to be enclosed by the same outer class. Only in that case was the outer field of the class omitted. We now omit if subclass is contained in an outer class that is itself a subclass of the superclasses outer class.

See test case "outertest.scala" for an example.

### CODE REVIEW

**Comment by odersky** [comment]
Review by @paulp

File: src/compiler/scala/tools/nsc/ast/TreeBrowsers.scala [hunk 1]
```diff
@@ -140,7 +140,7 @@ abstract class TreeBrowsers {
       UIManager.setLookAndFeel("com.sun.java.swing.plaf.nimbus.NimbusLookAndFeel")
     }
     catch {
-      case _ => UIManager.setLookAndFeel(UIManager.getCrossPlatformLookAndFeelClassName())
+      case _: Throwable => UIManager.setLookAndFeel(UIManager.getCrossPlatformLookAndFeelClassName())
```
Comment (viktorklang) [inline src/compiler/scala/tools/nsc/ast/TreeBrowsers.scala]: Should this really swallow OOMEs etc? Consider using scala.util.control.NonFatal

File: src/compiler/scala/tools/nsc/doc/model/ModelFactoryImplicitSupport.scala [hunk 2]
```diff
@@ -242,7 +242,7 @@ trait ModelFactoryImplicitSupport {
 
           available = Some(search.tree != EmptyTree)
         } catch {
-          case _ =>
+          case _ => // !!! [Martin]: One should never silently catch all Throwables. Please sharpen the exception type.
```
Comment (viktorklang) [inline src/compiler/scala/tools/nsc/doc/model/ModelFactoryImplicitSupport.scala]: scala.util.control.NonFatal here too?

File: src/compiler/scala/tools/nsc/util/InterruptReq.scala [hunk 3]
```diff
@@ -19,7 +19,7 @@ abstract class InterruptReq {
     try {
       result = Some(Left(todo()))
     } catch {
-      case t => result = Some(Right(t))
+      case t: Throwable => result = Some(Right(t))
     }
     notify()
```
Comment (viktorklang) [inline src/compiler/scala/tools/nsc/util/InterruptReq.scala]: Put the notify in a finally block?

**Comment by adriaanm** [comment]
@odersky, could you please update this pull request to not include the Throwable warning suppression (huge thanks for that, though)

here's the command to do that (assuming origin is your own github repository):

```
git push -f origin e0853b3255:optimize/outer
```

it says: push (even if it's not a fast-forward) to origin and set the branch optimize/outer to point to e0853b3255

**Comment by adriaanm** [comment]
(with the commit duplicated in #955 we'd get conflicts later)

**Comment by adriaanm** [comment]
to drop the commit in your own branch locally

```
git checkout optimize/outer
git log e0853b3255.. # verify that this indeed only lists the throwable commit
git reset --hard e0853b3255
```

**Comment by odersky** [comment]
Done. I was not aware that later pushes to a pullrequest affect the
pull request (did the pullrequest first, then pushed another commit to
optimize/outer, thinking that it would not affect the pull request.
Now I know better).

Thanks
- Martin

On Fri, Jul 20, 2012 at 5:03 PM, Adriaan Moors
reply@reply.github.com
wrote:

> to drop the commit in your own branch locally
> 
> ```
> git checkout optimize/outer
> git log e0853b3255.. # verify that this indeed only lists the throwable commit
> git reset --hard e0853b3255
> ```
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/scala/scala/pull/954#issuecomment-7132419

## 

Martin Odersky
Prof., EPFL and Chairman, Typesafe
PSED, 1015 Lausanne, Switzerland
Tel. EPFL: +41 21 693 6863
Tel. Typesafe: +41 21 691 4967

**Comment by paulp** [comment]
Hmm, somehow I missed the fact that this is closed.  In that case I'll just update it the old fashioned way.

**Comment by retronym** [comment]
You would need to push to optimize/outer on odersky/scala to add to the existing PR.

**Pull request merged**